File: sdk/typescript/src/decorators.ts
/**
 * IAOEP 高阶函数 — 装饰 Agent/LLM/Tool 调用产生 OTLP Span.
 */

import type { Tracer, Span } from '@opentelemetry/api';
import { SpanStatusCode } from '@opentelemetry/api';
import { getTracer } from './config.js';
import { randomUUID } from 'node:crypto';

// ============================================================
// 类型定义
// ============================================================

export interface ObservedAgentOptions {
  /** Agent 名称 (必填) */
  name: string;
  /** Skill 名 (可选) */
  skill?: string;
  /** 是否捕获输入参数 (PII 风险) */
  captureInput?: boolean;
  /** 是否捕获返回值 (PII 风险) */
  captureOutput?: boolean;
  /** payload 最大长度 (字符) */
  maxPayloadLength?: number;
}

export interface ObservedLLMOptions {
  /** LLM 供应商 (openai / dashscope / anthropic) */
  system: string;
  /** 从参数取模型名的路径 (e.g. "modelName" 或 "options.model") */
  modelPath?: string;
  /** 从参数取 input tokens 的路径 */
  inputTokensPath?: string;
  /** 从参数取 output tokens 的路径 */
  outputTokensPath?: string;
  /** 从返回值取 input tokens 的路径 (e.g. "usage.prompt_tokens") */
  inputTokensResultPath?: string;
  /** 从返回值取 output tokens 的路径 (e.g. "usage.completion_tokens") */
  outputTokensResultPath?: string;
}

export interface ObservedToolOptions {
  /** 工具名 (必填) */
  name: string;
  /** 是否捕获参数 */
  captureArguments?: boolean;
}

// ============================================================
// 工具
// ============================================================

function truncate(s: string, max: number): string {
  return s.length <= max ? s : s.substring(0, max) + '...';
}

function safeJson(obj: unknown): string {
  try {
    return JSON.stringify(obj, (_k, v) => v ?? null);
  } catch {
    return String(obj);
  }
}

function resolvePath(obj: unknown, path: string): unknown {
  if (!obj || !path) return undefined;
  const parts = path.split('.');
  let cur: unknown = obj;
  for (const part of parts) {
    if (cur === null || cur === undefined) return undefined;
    cur = (cur as Record<string, unknown>)[part];
  }
  return cur;
}

async function runWithSpan<F extends (...args: never[]) => unknown>(
  tracer: Tracer,
  spanName: string,
  setAttributes: (span: Span) => void,
  func: F,
  args: Parameters<F>,
): Promise<ReturnType<F>> {
  const span = tracer.startSpan(spanName);
  setAttributes(span);
  try {
    const result = await func(...args);
    span.setStatus({ code: SpanStatusCode.OK });
    return result as ReturnType<F>;
  } catch (e) {
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: e instanceof Error ? e.message : String(e),
    });
    span.recordException(e as Error);
    throw e;
  } finally {
    span.end();
  }
}

// ============================================================
// observedAgent
// ============================================================

export function observedAgent<TArgs extends unknown[], TReturn>(
  options: ObservedAgentOptions,
  func: (...args: TArgs) => TReturn | Promise<TReturn>,
): (...args: TArgs) => Promise<TReturn> {
  const tracer = getTracer();
  return async (...args: TArgs): Promise<TReturn> => {
    return runWithSpan(
      tracer,
      'agent.run',
      (span) => {
        span.setAttribute('agent.name', options.name);
        if (options.skill) {
          span.setAttribute('agent.skill.name', options.skill);
        }
        if (options.captureInput) {
          span.setAttribute(
            'agent.input',
            truncate(safeJson(args), options.maxPayloadLength ?? 4096),
          );
        }
      },
      func,
      args,
    );
  };
}

// ============================================================
// observedLLMCall
// ============================================================

export function observedLLMCall<TArgs extends unknown[], TReturn>(
  options: ObservedLLMOptions,
  func: (...args: TArgs) => TReturn | Promise<TReturn>,
): (...args: TArgs) => Promise<TReturn> {
  const tracer = getTracer();
  return async (...args: TArgs): Promise<TReturn> => {
    return runWithSpan(
      tracer,
      'llm.call',
      (span) => {
        span.setAttribute('gen_ai.system', options.system);

        // 模型名: 从参数取
        if (options.modelPath) {
          const modelVal = resolvePath(args[0], options.modelPath);
          if (modelVal !== undefined) {
            span.setAttribute('gen_ai.request.model', String(modelVal));
          }
        }

        // Token 数: 从参数取
        if (options.inputTokensPath) {
          const val = resolvePath(args[0], options.inputTokensPath);
          if (val !== undefined) {
            span.setAttribute('gen_ai.usage.input_tokens', Number(val));
          }
        }
        if (options.outputTokensPath) {
          const val = resolvePath(args[0], options.outputTokensPath);
          if (val !== undefined) {
            span.setAttribute('gen_ai.usage.output_tokens', Number(val));
          }
        }
      },
      func,
      args,
    );
  };
}

// ============================================================
// observedTool
// ============================================================

export function observedTool<TArgs extends unknown[], TReturn>(
  options: ObservedToolOptions,
  func: (...args: TArgs) => TReturn | Promise<TReturn>,
): (...args: TArgs) => Promise<TReturn> {
  const tracer = getTracer();
  return async (...args: TArgs): Promise<TReturn> => {
    return runWithSpan(
      tracer,
      'tool.execute',
      (span) => {
        span.setAttribute('tool.name', options.name);
        span.setAttribute('tool.call.id', randomUUID());
        if (options.captureArguments) {
          span.setAttribute('tool.arguments', safeJson(args));
        }
      },
      func,
      args,
    );
  };
}

File: sdk/go/decorators.go
package iaoep

import (
	"context"
	"crypto/rand"
	"fmt"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/trace"
)

// ObservedAgent 装饰一个 Agent 方法, 自动产生 agent.run Span.
//
// 用法:
//
//	result, err := iaoep.ObservedAgent(ctx, "customer-service", "general",
//	    func(ctx context.Context) (string, error) {
//	        return callLLM(ctx, query)
//	    })
//
// 参数:
//   - name:   Agent 名 (必填, 作为 agent.name 属性)
//   - skill:  Skill 名 (可选)
//   - fn:     实际函数
func ObservedAgent[T any](ctx context.Context, name, skill string, fn func(context.Context) (T, error)) (T, error) {
	tracer := Tracer().(trace.Tracer)
	ctx, span := tracer.Start(ctx, "agent.run",
		trace.WithSpanKind(trace.SpanKindInternal),
	)
	defer span.End()

	span.SetAttributes(attribute.String("agent.name", name))
	if skill != "" {
		span.SetAttributes(attribute.String("agent.skill.name", skill))
	}

	result, err := fn(ctx)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return result, err
	}
	span.SetStatus(codes.Ok, "")
	return result, nil
}

// ObservedLLMCall 装饰一个 LLM 调用方法, 产生 llm.call Span.
//
// 用法:
//
//	resp, err := iaoep.ObservedLLMCall(ctx, "openai", "qwen3-turbo", 120, 80,
//	    func(ctx context.Context) (string, error) {
//	        return openaiClient.Chat(model, messages)
//	    })
//
// 参数:
//   - system:        LLM 供应商 (openai / dashscope / glm / deepseek)
//   - model:        模型名
//   - inputTokens:  输入 token 数 (可选, 0 = 跳过)
//   - outputTokens: 输出 token 数 (可选, 0 = 跳过)
//   - fn:            实际 LLM 调用
func ObservedLLMCall[T any](ctx context.Context, system, model string,
	inputTokens, outputTokens int,
	fn func(context.Context) (T, error)) (T, error) {

	tracer := Tracer().(trace.Tracer)
	ctx, span := tracer.Start(ctx, "llm.call",
		trace.WithSpanKind(trace.SpanKindClient),
	)
	defer span.End()

	span.SetAttributes(
		attribute.String("gen_ai.system", system),
		attribute.String("gen_ai.request.model", model),
	)
	if inputTokens > 0 {
		span.SetAttributes(attribute.Int("gen_ai.usage.input_tokens", inputTokens))
	}
	if outputTokens > 0 {
		span.SetAttributes(attribute.Int("gen_ai.usage.output_tokens", outputTokens))
	}

	result, err := fn(ctx)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return result, err
	}
	span.SetStatus(codes.Ok, "")
	return result, nil
}

// ObservedTool 装饰一个工具调用方法, 产生 tool.execute Span.
func ObservedTool[T any](ctx context.Context, name string,
	fn func(context.Context) (T, error)) (T, error) {

	tracer := Tracer().(trace.Tracer)
	ctx, span := tracer.Start(ctx, "tool.execute",
		trace.WithSpanKind(trace.SpanKindInternal),
	)
	defer span.End()

	span.SetAttributes(
		attribute.String("tool.name", name),
		attribute.String("tool.call.id", randomUUID()),
	)

	result, err := fn(ctx)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return result, err
	}
	span.SetStatus(codes.Ok, "")
	return result, nil
}

// ObservedABTest 装饰一个方法, 根据 trace_id 分配 baseline/candidate (Phase 4-5).
//
// 用法:
//
//	result, err := iaoep.ObservedABTest(ctx, "customer-service-ab",
//	    func(ctx context.Context) (string, error) {
//	        group := iaoep.ABTestGroupFromContext(ctx)
//	        // 根据 group 选不同实现
//	        return callLLM(ctx, modelForGroup(group))
//	    })
func ObservedABTest[T any](ctx context.Context, abTestName string,
	fn func(context.Context) (T, error)) (T, error) {

	group := "baseline"   // 简化: 实际从 evaluator API 调 /assign

	tracer := Tracer().(trace.Tracer)
	ctx, span := tracer.Start(ctx, "ab.test",
		trace.WithSpanKind(trace.SpanKindInternal),
	)
	defer span.End()

	span.SetAttributes(
		attribute.String("ab_test_name", abTestName),
		attribute.String("ab_test_group", group),
	)

	ctx = contextWithABGroup(ctx, group)
	result, err := fn(ctx)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return result, err
	}
	span.SetStatus(codes.Ok, "")
	return result, nil
}

// ABTestGroupFromContext 从 context 读当前 A/B Test group.
func ABTestGroupFromContext(ctx context.Context) string {
	if v := ctx.Value(abTestGroupKey); v != nil {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return "baseline"   // 兜底
}

// ---------- private helpers ----------

type contextKey string

const abTestGroupKey contextKey = "iaoep.ab_test_group"

func contextWithABGroup(ctx context.Context, group string) context.Context {
	return context.WithValue(ctx, abTestGroupKey, group)
}

// randomUUID 生成 UUID v4 (使用 crypto/rand).
func randomUUID() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant 10
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
